### PR TITLE
Implement EVM `BYTE` operation

### PR DESCRIPTION
**Pull request by unzvfu**
Like it says on the tin. Thanks for your help @nbgl! Your prize is you get to review. :)

### CODE REVIEW

File: evm/src/arithmetic/byte.rs [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+//! INPUTS: 256-bit values I and X = \sum_{i=0}^31 X_i B^i,
+//! where B = 2^8 and 0 <= X_i < B for all i.
+//!
+//! OUTPUT: X_{31-I} if 0 <= I < 32, otherwise 0.
+//!
+//! NB: index I=0 corresponds to byte X_31, i.e. the most significant
+//! byte. This is exactly the opposite of anyone would expect; who
+//! knows what the EVM designers were thinking. Anyway, if anything
+//! below seems confusing, first check to ensure you're counting from
+//! the wrong end of X, as the spec requires.
```
Comment (nbgl) [inline evm/src/arithmetic/byte.rs:14]: lol

File: evm/src/arithmetic/byte.rs [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+//! 2. y * 256 ∈ {0, 256, 512, ..., 2^24 - 512, 2^24 - 256}
+//! 3. w / 256 = L - y * 256 ∈ {-2^24 + 256, -2^24 + 257, ..., 2^16 - 2, 2^16 - 1}
+//! 4. By inspection, for w < 2^16, if w / 256 < 2^16 or
+//!    w / 256 >= P - 2^24 + 256 (i.e. if w / 256 falls in the range
+//!    of point 3 above), then w = 256 * m for some 0 <= m < 256.
+//! 5. Hence w / 256 ∈ {0, 1, ..., 255}
+//! 6. Hence y * 256 = L - w / 256 ∈ {-255, -254, ..., 2^16 - 1}
+//! 7. Taking the intersection of ranges in 2. and 6. we see that
+//!    y * 256 ∈ {0, 256, 512, ..., 2^16 - 256}
+//! 8. Hence y ∈ {0, 1, ..., 255}
```
Comment (nbgl) [inline evm/src/arithmetic/byte.rs:61]: Excellent explanation and a pleasure to read ☺️

File: evm/src/arithmetic/byte.rs [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+use static_assertions::const_assert;
+
+use crate::arithmetic::columns::*;
+use crate::arithmetic::utils::u256_to_array;
+use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+
+// Give meaningful names to the columns of AUX_INPUT_REGISTER_0 that
+// we're using
+const BYTE_LAST_LIMB_LO: usize = AUX_INPUT_REGISTER_0.start + 7;
```
Comment (nbgl) [inline evm/src/arithmetic/byte.rs:78]: Why start at `AUX_INPUT_REGISTER_0.start + 7` when `AUX_INPUT_REGISTER_0.start + 6` is unused?

File: evm/src/arithmetic/byte.rs [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+/// the higher 11 bits of the first limb, so we put them in
+/// `idx_decomp[5]`. The rest of `idx_decomp` is set to 0.
+fn set_idx_decomp<F: PrimeField64>(idx_decomp: &mut [F], idx: &U256) {
+    for i in 0..5 {
+        idx_decomp[i] = F::from_bool(idx.bit(i));
+    }
+    idx_decomp[5] = F::from_canonical_u16((idx.low_u64() as u16) >> 5);
+    for i in 6..N_LIMBS {
+        idx_decomp[i] = F::ZERO;
+    }
```
Comment (nbgl) [inline evm/src/arithmetic/byte.rs:103]: The issue with zero-filling the rest of this register is that it's shared with `BYTE_LAST_LIMB_LO` and friends. If the next maintainer writes to `BYTE_LAST_LIMB_LO` and then calls this seemingly unrelated function, `BYTE_LAST_LIMB_LO` will be erased and our future pal will be confused.

File: evm/src/arithmetic/byte.rs [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+}
```
Comment (nbgl) [inline evm/src/arithmetic/byte.rs:104]: I would make this function accept a slice of length 6, not of length `N_LIMBS`. That way you don't have to worry about the remaining limbs.

File: evm/src/arithmetic/byte.rs [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+
+    let (prev, next) = next.split_at_mut(lvl_len);
+    let lvl_len = 2;
+    let offset = (!idx.bit(2) as usize) * lvl_len;
+    next[..lvl_len].copy_from_slice(&prev[offset..offset + lvl_len]);
+
+    let (prev, next) = next.split_at_mut(lvl_len);
+    // lvl_len = 1
+    let offset = !idx.bit(1) as usize;
+    next[0] = prev[offset]; // tree[14] = tree[offset + 12]
```
Comment (nbgl) [inline evm/src/arithmetic/byte.rs:182]: How about [`copy_within`](https://doc.rust-lang.org/std/primitive.slice.html#method.copy_within)?

File: evm/src/arithmetic/byte.rs [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+    // low 5 bits of the first limb of idx:
+    let mut idx0_lo5 = P::ZEROS;
+    for i in 0..5 {
+        let bit = idx_decomp[i];
+        yield_constr.constraint(is_byte * (bit * bit - bit));
+        idx0_lo5 += bit * P::Scalar::from_canonical_u64(1 << i);
+    }
+    // high (11) bits of the first limb of idx:
+    let idx0_hi = idx_decomp[5] * P::Scalar::from_canonical_u64(32u64);
+    yield_constr.constraint(is_byte * (idx[0] - (idx0_lo5 + idx0_hi)));
```
Comment (nbgl) [inline evm/src/arithmetic/byte.rs:233]: I would briefly explain that this guarantees that `idx0_hi` is range-checked to 11-bits.

File: evm/src/arithmetic/byte.rs [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+    debug_assert!(base8 * base8_inv == P::Scalar::ONE);
+
+    let lo_byte = lv[BYTE_LAST_LIMB_LO];
+    let hi_byte = lv[BYTE_LAST_LIMB_HI];
+    yield_constr.constraint(is_byte * (base8_inv * lo_byte + base8 * hi_byte - limb));
+
+    let bit = idx_decomp[0];
+    let t = bit * base8_inv * lo_byte + (P::ONES - bit) * hi_byte;
+    yield_constr.constraint(is_byte * (tree[15] - t));
+    let expected_out_byte = tree[15];
```
Comment (nbgl) [inline evm/src/arithmetic/byte.rs:275]: To make the constraints field-agnostic but still avoiding doing inverses, we can rewrite the constraints to remove `base8_inv`. Just multiply each constraint by `base8`.

Something like:
```suggestion
    let base8 = P::Scalar::from_canonical_u64(1 << 8);
    // base8_inv is Goldilocks-specific
    debug_assert!(P::Scalar::order() == 0xffff_ffff_0000_0001u64.into());
    let base8_inv = P::Scalar::from_canonical_u64(0xfeffffff01000001);
    debug_assert!(base8 * base8_inv == P::Scalar::ONE);

    let lo_byte = lv[BYTE_LAST_LIMB_LO];
    let hi_byte = lv[BYTE_LAST_LIMB_HI];
    yield_constr.constraint(is_byte * (lo_byte + base8 * (base8 * hi_byte - limb)));

    let bit = idx_decomp[0];
    let t = bit * lo_byte + (P::ONES - bit) * base8 * hi_byte;
    yield_constr.constraint(is_byte * (base8 * tree[15] - t));
    let expected_out_byte = tree[15];
```

File: evm/src/arithmetic/byte.rs [hunk 9]
```diff
@@ -0,0 +1,502 @@
...
+
+    // Sum all higher limbs; sum will be non-zero iff idx >= 32.
+    //let hi_limb_sum: P = idx0_hi + idx[1..].iter().sum::<P>(); // doesn't work
+    let hi_limb_sum = idx[1..].iter().fold(idx0_hi, |acc, &i| i + acc);
```
Comment (nbgl) [inline evm/src/arithmetic/byte.rs:279]: I don't think packed addition is implemented for references, but an explicit copy should fix it:

```suggestion
    let hi_limb_sum: P = idx0_hi + idx[1..].iter().copied().sum::<P>();
```

**Review by nbgl — APPROVED** [review]
Clean code and fantastic documentation. I pointed out a few minor places for improvement but no blockers :)

File: evm/src/arithmetic/byte.rs (hunk 3 above)
Comment (unzvfu) [reply evm/src/arithmetic/byte.rs:78]: Because counting is haaaaard. :)

File: evm/src/arithmetic/byte.rs (hunk 6 above)
Comment (unzvfu) [reply evm/src/arithmetic/byte.rs:182]: Haha, StackOverflow led me astray! Agreed, that's much cleaner.

**Pull request merged**